backend/api/communication/src/main/java/co/airy/core/api/communication/AsyncSendMessagesHandler.java

```diff
@@ -0,0 +1,159 @@
+import org.springframework.stereotype.Component;
+import org.slf4j.Logger;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+import static co.airy.model.metadata.MetadataRepository.newMessageMetadata;
+
+@Component
+public class AsyncSendMessagesHandler implements Runnable {
+
+    private static final Logger log = AiryLoggerFactory.getLogger(AsyncSendMessagesHandler.class);
+
+    @AllArgsConstructor
+    @Data
+    private class PendingMessage {
+        private Message msg;
+        private Instant now;
+    }
+
+    private final Stores stores;
+    private final ConcurrentLinkedQueue<PendingMessage> pendingConversations;
+    private final Thread thread;
+    private final long maxWaitMillis;
+    private final long periodMillis;
+    private volatile boolean keepAlive;
+    
+
+    AsyncSendMessagesHandler(
+            Stores stores,
+            @Value("${async-messages-handler.max-wait-millis}")
+            long maxWaitSeconds,
+            @Min(1)
+            @Value("${async-messages-handler.check-period-millis}")
+            long periodMillis) {
+        this.stores = stores;
+        this.pendingConversations = new ConcurrentLinkedQueue<>();
+        this.keepAlive = true;
+        this.maxWaitMillis = maxWaitSeconds;
+        this.periodMillis = periodMillis;
+        this.thread = new Thread(this);
+    }
+
+    @Override
+    public void run() {
+        final List<PendingMessage> pending = new LinkedList<>();
+
+        while (keepAlive) {
+            try {
+                Thread.sleep(this.periodMillis);
+
+                // Move all values to internal list
+                PendingMessage pm;
+                while ((pm = pendingConversations.poll()) != null) {
+                    pending.add(pm);
+                }
+
+                final ListIterator<PendingMessage> iter = pending.listIterator();
+                while (iter.hasNext()) {
+                    PendingMessage pendingMsg = iter.next();
+
+
+                    final ReadOnlyKeyValueStore<String, Conversation> conversationsStore = stores.getConversationsStore();
+                    final Conversation conversation = conversationsStore.get(pendingMsg.getMsg().getConversationId());
+                    final boolean messageExpired = Duration.between(pendingMsg.getNow(), Instant.now()).toMillis() > maxWaitMillis;
+
+                    if (conversation == null && !messageExpired) {
+                        continue;
+                    }
+
+                    Message msg = pendingMsg.getMsg();
+
+                    if (conversation != null) {
+                        Channel channel = conversation.getChannel();
+                        if (channel.getConnectionState().equals(ChannelConnectionState.DISCONNECTED)) {
+                            msg = setMessageSateToFailed(msg, "Channel not connected");
+                        }
+                        updateMessage(conversation, msg);
+
+                    } else if (messageExpired) {
```

Comment:
`messageExpired` is always true because of the if check in line 92

Reply:
It is inside a `else if` because it can be a case where the message expired but the conversation is there. so in that case we just want to ingest the message.

Reply:
That's true, but what I'm saying is that if the conversation is not null, and the else gets triggered the message has to be expired because of the check. The `else if` is redundant.